### ISSUE TITLE
Travis CI - test remote coderunner - travis_ci branch

### ISSUE THREAD

**Issue by ferdinandvanwyk** [edmundhighcock/coderunner#6]
I'm trying to get automated testing set up on the Travis CI service. In short it's a clean copy of Ubuntu 12.04 LTS in a container in which I have sudo access.

In order to make it work, you write a recipe of things which need to be installed/compiled before the tests can be run, expressed as a `.travis.yml` file: https://github.com/coderunner-framework/coderunner/blob/travis_ci/.travis.yml. It should be fairly self-explanatory: just a series of bash commands to get the environment set up.

The problem is occurring with testing remote CR. In particular:

``` bash
...
testing remote coderunner
Calling Remote Method: marshalled_variables
/home/travis/build/coderunner-framework/coderunner/lib/coderunner/remote_code_runner.rb:152: warning: instance variable @remote_cache not initialized
Connecting to server using 'ssh travis@localhost'...
Loading folder /home/travis/build/coderunner-framework/coderunner/test/results...
cd /home/travis/build/coderunner-framework/coderunner/test/results
export ROWS=24
export COLS=80 
source /etc/bashrc /etc/profile > /dev/null 2> /dev/null
source ~/.bashrc > /dev/null 2> /dev/null  ; source ~/.bash_login > /dev/null 2> /dev/null  ; source ~/.bash_profile > /dev/null 2> /dev/null  ; source ~/.profile > /dev/null 2> /dev/null 
if [ "$CODE_RUNNER_COMMAND" ]
    then
        $CODE_RUNNER_COMMAND runner_eval ";marshalled_variables(*[])" -Z "{:U=>true, :N=>true, :G=>[], :g=>[], :e=>nil, :r=>nil, :w=>nil, :Y=>nil}" 
    else
        coderunner runner_eval ";marshalled_variables(*[])" -Z "{:U=>true, :N=>true, :G=>[], :g=>[], :e=>nil, :r=>nil, :w=>nil, :Y=>nil}"
fi

bash: line 9: coderunner: command not found
...
```

The full log can be found here: https://s3.amazonaws.com/archive.travis-ci.org/jobs/120265409/log.txt

It's saying that the `coderunner` command cannot be found on the 'remote' system after SSH'ing into the `localhost`, but `gem install coderunner` was run as part of the `.travis.yml` setup.

Any ideas why it can't find the `coderunner` command?

**Comment by ferdinandvanwyk** [comment]
I should add that it leads to the following failed test:

``` bash
1) Error:
TestCodeRunner#test_remote_coderunner:
NoMethodError: undefined method `[]' for nil:NilClass
    /home/travis/build/coderunner-framework/coderunner/lib/coderunner/remote_code_runner.rb:246:in `update'
    /home/travis/build/coderunner-framework/coderunner/test/test_coderunner.rb:128:in `test_remote_coderunner'
```

**Comment by edmundhighcock** [comment]
When gem install coderunner installs it on the test system, where does it install it to? Is coderunner definitely in the PATH when RemoteCodeRunner logs in to localhost via ssh?

**Comment by edmundhighcock** [comment]
You have this line in preinstall

export PATH=$PWD/tools/bin:$PATH

 but this is not being put in the bashrc

**Comment by ferdinandvanwyk** [comment]
Good catch, thanks. Let's see what it says. Currently testing take over 10min since I'm sticking to compiling from scratch. Once I have something that passes the tests I'll work on optimizing them with packages installed through the OS package manager.

p.s. It's using RVM so gems are install to the default directory RVM uses.

**Comment by ferdinandvanwyk** [comment]
This particular part of the testing process turned into a headache, but I solved it by adding an additional source command when connecting to a remote machine via ssh. When connecting to the localhost on Travis CI, the environment wouldn't necessarily be the same in particular RVM wasn't being sourced properly. The fact that you're not using RVM on other machines shouldn't be a problem since you're already sourcing things which may or may not be there.

See PR #8 which passes all the tests and is ready to merge. I mentioned the commit squashing feature but I can't seem to make this work right now so I'm going to merge this as is (with several commits where I just had to play around and test the Travis environment)
